### PR TITLE
use knp voter for the active menu

### PR DESCRIPTION
**Pull request by andrey1s**

### CODE REVIEW

**Comment by core23** [comment]
Can you say something more?

**Comment by andrey1s** [comment]
@core23 the [KnpMenu](https://github.com/KnpLabs/KnpMenu/blob/master/doc/01-Basic-Menus.markdown#the-current-menu-item) uses several Voters to determine the active menu item, I just wrote the Voters to use with Sonata

File: Menu/Matcher/Voter/AdminVoter.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+namespace Sonata\AdminBundle\Menu\Matcher\Voter;
+
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Matcher\Voter\VoterInterface;
+use Sonata\AdminBundle\Admin\AdminInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+class AdminVoter implements VoterInterface
+{
+    /** @var Request */
```
Comment (OskarStark) [inline Menu/Matcher/Voter/AdminVoter.php]: please use 3 lines

File: Menu/Matcher/Voter/ChildrenVoter.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace Sonata\AdminBundle\Menu\Matcher\Voter;
+
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Matcher\MatcherInterface;
+use Knp\Menu\Matcher\Voter\VoterInterface;
+
+class ChildrenVoter implements VoterInterface
+{
+    /** @var MatcherInterface */
```
Comment (OskarStark) [inline Menu/Matcher/Voter/ChildrenVoter.php]: please use 3 lines

File: Resources/config/menu.xml [hunk 3]
```diff
@@ -16,5 +16,18 @@
         <service id="sonata.admin.sidebar_menu" class="Knp\Menu\MenuItem">
             <tag name="knp_menu.menu" alias="sonata_admin_sidebar" />
         </service>
+
+        <service id="sonata.admin.menu.mather.voter.admin" class="Sonata\AdminBundle\Menu\Matcher\Voter\AdminVoter">
```
Comment (OskarStark) [inline Resources/config/menu.xml]: `mather` or `matcher`??

File: Resources/views/Menu/sonata_menu.html.twig [hunk 4]
```diff
@@ -16,31 +16,8 @@
...
-
-        {%- if item.hasChildren %}
-            {%- do item.setAttribute('class', (item.attribute('class')~' treeview')|trim) %}
-        {%- endif %}
-        {%- if active %}
-            {%- do item.setAttribute('class', (item.attribute('class')~' active')|trim) %}
-            {%- do item.setChildrenAttribute('class', (item.childrenAttribute('class')~' active')|trim) %}
-        {%- endif %}
-
+        {% set options = options|merge({branch_class:'treeview',currentClass:"active"}) %}
```
Comment (OskarStark) [inline Resources/views/Menu/sonata_menu.html.twig]: please add spaces to the options array

File: Tests/Menu/Matcher/Voter/AbstractVoterTest.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+abstract class AbstractVoterTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @return array
+     */
+    abstract public function provideData();
+
+    /**
+     */
```
Comment (OskarStark) [inline Tests/Menu/Matcher/Voter/AbstractVoterTest.php]: please fix the php doc

**Comment by OskarStark** [comment]
thank you for your work @andrey1s , i like your work and the tests :+1: 

@core23 i think we can merge this PR when my comments are applied

File: Resources/views/Menu/sonata_menu.html.twig [hunk 6]
```diff
@@ -16,31 +16,8 @@
...
-        {%- if item.hasChildren %}
-            {%- do item.setAttribute('class', (item.attribute('class')~' treeview')|trim) %}
-        {%- endif %}
-        {%- if active %}
-            {%- do item.setAttribute('class', (item.attribute('class')~' active')|trim) %}
-            {%- do item.setChildrenAttribute('class', (item.childrenAttribute('class')~' active')|trim) %}
-        {%- endif %}
-
+        {% set options = options|merge({branch_class: 'treeview', currentClass: "active"}) %}
+        {%- do item.setChildrenAttribute('class', (item.childrenAttribute('class')~' active')|trim) %}
```
Comment (greg0ire) [inline Resources/views/Menu/sonata_menu.html.twig]: Wow. Nice simplification!

File: Tests/Menu/Matcher/Voter/AbstractVoterTest.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * @param mixed $data
+     *
+     * @return ItemInterface
+     */
+    abstract protected function createItem($data);
+
+    /**
+     * @param mixed     $dataItem
+     * @param mixed     $dataVoter
```
Comment (greg0ire) [inline Tests/Menu/Matcher/Voter/AbstractVoterTest.php]: `itemData`, `voterData` would be a bit better IMO

File: Tests/Menu/Matcher/Voter/ActiveVoterTest.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+     * @return ItemInterface
+     */
+    protected function createItem($data)
+    {
+        $item = $this->getMock('Knp\Menu\ItemInterface');
+        $item->expects($this->any())
+             ->method('getExtra')
+             ->with('active')
+             ->will($this->returnValue($data))
+        ;
```
Comment (greg0ire) [inline Tests/Menu/Matcher/Voter/ActiveVoterTest.php]: Using prophecy would make this kind of test easier to read: 

```
$item = $this->prophesize('Knp\Menu\ItemInterface');
$item->getExtra('active')->willReturn($data);

return $item->reveal();
```

No need to change the test though, it is just a tip for your next PR ;)

File: Tests/Menu/Matcher/Voter/AbstractVoterTest.php (hunk 7 above)
Comment (OskarStark) [inline Tests/Menu/Matcher/Voter/AbstractVoterTest.php]: :+1:

File: Resources/views/Menu/sonata_menu.html.twig (hunk 6 above)
Comment (OskarStark) [inline Resources/views/Menu/sonata_menu.html.twig]: if it works :smile: .. then yes :+1:

Comment (hason) [inline Resources/views/Menu/sonata_menu.html.twig]: Duplication of #3065

**Comment by core23** [comment]
Ping @OskarStark

**Comment by greg0ire** [comment]
@hason there are conflicts with #3149 , can you resolve them ?

**Comment by OskarStark** [comment]
@greg0ire why are you pinging @hason, because this PR is done by @andrey1s 

IMO we need to decide if we would like to merge this one or this one: #3065

ping @sonata-project/contributors

**Comment by greg0ire** [comment]
> @greg0ire why are you pinging @hason, because this PR is done by @andrey1s 

Woops, sorry @hason , I was mistaken

**Comment by greg0ire** [comment]
> IMO we need to decide if we would like to merge this one or this one: #3065

I'd say this one : it has tests

**Comment by core23** [comment]
:+1: for this PR

Maybe some cleanup from the other PR could be applied too?

File: Menu/Matcher/Voter/ActiveVoter.php [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Menu\Matcher\Voter;
+
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Matcher\Voter\VoterInterface;
+
```
Comment (Soullivaneuh) [inline Menu/Matcher/Voter/ActiveVoter.php]: Please add class PHP doc with `@author` tag and if relevant, a description.

File: Menu/Matcher/Voter/AdminVoter.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Menu\Matcher\Voter;
+
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Matcher\Voter\VoterInterface;
+use Sonata\AdminBundle\Admin\AdminInterface;
+use Symfony\Component\HttpFoundation\Request;
+
```
Comment (Soullivaneuh) [inline Menu/Matcher/Voter/AdminVoter.php]: Please add class PHP doc with `@author` tag and if relevant, a description.

File: Menu/Matcher/Voter/ChildrenVoter.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Menu\Matcher\Voter;
+
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Matcher\MatcherInterface;
+use Knp\Menu\Matcher\Voter\VoterInterface;
+
```
Comment (Soullivaneuh) [inline Menu/Matcher/Voter/ChildrenVoter.php]: Please add class PHP doc with `@author` tag and if relevant, a description.

**Comment by core23** [comment]
Can you have a look at this @andrey1s

**Comment by andrey1s** [comment]
@core23 rebased, added author and small description

**Comment by OskarStark** [comment]
Thank you @andrey1s 👍

**Comment by Soullivaneuh** [comment]
@andrey1s Your PR introduce this error:

```
Symfony\Component\DependencyInjection\Exception\ServiceCircularReferenceException: Circular reference detected for service "sonata.admin.menu.matcher.voter.children", path: "sonata.admin.menu.matcher.voter.children -> knp_menu.matcher".
```

Could you please take a look?

**Comment by andrey1s** [comment]
@Soullivaneuh I can not repeat, you can more error conditions, package versions, and so. So I can repeat.

**Comment by Soullivaneuh** [comment]
@andrey1s Maybe my installed packages list could help:

```bash
$ composer show sonata-project/*
sonata-project/admin-bundle              3.0.0  The missing Symfony Admin Generator
sonata-project/block-bundle              3.0.0  Symfony SonataBlockBundle
sonata-project/cache                     1.0.7  Cache library
sonata-project/core-bundle               3.0.0  Symfony SonataCoreBundle
sonata-project/doctrine-extensions       1.0.2  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.0.0  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.1.10 Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.4.1  Lightweight Exporter library
sonata-project/intl-bundle               2.2.4  Symfony SonataIntlBundle
$ composer show symfony/*
symfony/assetic-bundle     v2.8.0  Integrates Assetic into Symfony2
symfony/monolog-bundle     2.11.1  Symfony MonologBundle
symfony/polyfill-apcu      v1.1.1  Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu  v1.1.1  Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.1.1  Symfony polyfill for the Mbstring extension
symfony/polyfill-php54     v1.1.1  Symfony polyfill backporting some PHP 5.4+ features to lower PHP versions
symfony/polyfill-php55     v1.1.1  Symfony polyfill backporting some PHP 5.5+ features to lower PHP versions
symfony/polyfill-php56     v1.1.1  Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.1.1  Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.1.1  Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.0  Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.3.11 Symfony SwiftmailerBundle
symfony/symfony            v2.8.6  The Symfony PHP framework
$ composer show knplabs/*
knplabs/knp-components       1.3.3  Knplabs component library
knplabs/knp-markdown-bundle  1.5.1  Knplabs markdown bundle transforms markdown into html
knplabs/knp-menu             v2.1.1 An object oriented menu library
knplabs/knp-menu-bundle      v2.1.1 This bundle provides an integration of the KnpMenu library
knplabs/knp-paginator-bundle 2.5.3  Paginator bundle for Symfony2 to automate pagination and simplify sorting and other features
```

**Comment by greg0ire** [comment]
@Soullivaneuh maybe give your composer.json ?

**Comment by Soullivaneuh** [comment]
@greg0ire Prepare your heart attack. :heart: 

```json
{
    "autoload": {
        "psr-4": { "": "src/" },
        "files": [ "app/AppKernel.php" ]
    },
    "autoload-dev": {
        "psr-4": { "Tests\\": "tests/" }
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/Soullivaneuh/dompdf.git"
        },
        {
            "type": "vcs",
            "url": "https://github.com/nexylan/NelmioApiDocBundle.git"
        },
        {
            "type": "vcs",
            "url": "https://github.com/nexylan/WebApiExtension.git"
        }
    ],
    "require": {
        "php": "^5.6",
        "ext-intl": ">=1.0.2",
        "symfony/symfony": "2.8.*",
        "doctrine/orm": "^2.4",
        "doctrine/doctrine-bundle": "^1.4",
        "doctrine/mongodb-odm-bundle": "^3.0",
        "doctrine/doctrine-migrations-bundle": "^1.0",
        "stof/doctrine-extensions-bundle": "^1.1",
        "twig/extensions": "^1.0",
        "symfony/assetic-bundle": "^2.7",
        "symfony/swiftmailer-bundle": "^2.3",
        "symfony/monolog-bundle": "^2.4",
        "sensio/distribution-bundle": "^5.0",
        "sensio/framework-extra-bundle": "^3.0.8",
        "sensio/buzz-bundle": "^1.0",
        "incenteev/composer-parameter-handler": "^2.0",
        "jms/serializer-bundle": "^1.0",
        "gregwar/image-bundle": "^2.0",
        "genemu/form-bundle": "^2.3",
        "mopa/bootstrap-bundle": "^3.0",
        "knplabs/knp-paginator-bundle": "^2.4.3",
        "knplabs/knp-menu-bundle": "^2.0",
        "knplabs/knp-markdown-bundle": "^1.3",
        "lexik/paybox-bundle": "^2.0",
        "lexik/jwt-authentication-bundle": "^1.0",
        "whiteoctober/breadcrumbs-bundle": "^1.1",
        "dompdf/dompdf": "dev-bs3-fix",
        "friendsofsymfony/user-bundle": "^2.0",
        "friendsofsymfony/oauth-server-bundle": "^1.4",
        "friendsofsymfony/rest-bundle": "^1.5",
        "friendsofsymfony/jsrouting-bundle": "^1.5",
        "nelmio/api-doc-bundle": "dev-nexy as 2.10",
        "sonata-project/admin-bundle": "^3.0",
        "sonata-project/doctrine-orm-admin-bundle": "^3.0",
        "sonata-project/intl-bundle": "^2.2",
        "sonata-project/doctrine-extensions": "^1.0",
        "ornicar/gravatar-bundle": "^1.1.3",
        "fabpot/goutte": "^3.0",
        "digitick/sepa-xml" : "^1.0",
        "guzzlehttp/guzzle": "^6.0",
        "rollerworks/password-strength-bundle": "^1.0",
        "hackzilla/password-generator-bundle": "^2.0",
        "a2lix/translation-form-bundle": "^1.3",
        "kriswallsmith/spork": "^0.3",
        "cebe/markdown": "^1.1",
        "browscap/browscap-bundle": "^1.0",
        "inlinestyle/inlinestyle": "^1.2",
        "hautelook/alice-bundle": "^1.0",
        "greg0ire/enum": "^2.1",
        "sllh/iso-codes-validator": "^1.1",
        "nexylan/slack-bundle": "^1.0",
        "nexylan/paybox-direct": "^0.2",
        "nexylan/cloudflare": "dev-master",
        "nexylan/cloudflare-host-gw": "dev-master",
        "nexylan/nexycrypt": "dev-master"
    },
    "require-dev": {
        "sensio/generator-bundle": "^3.0",
        "sonata-project/easy-extends-bundle": "^2.1",
        "doctrine/doctrine-fixtures-bundle": "^2.1",
        "behat/symfony2-extension": "^2.0",
        "behat/web-api-extension": "dev-nexy as 1.0",
        "behat/mink-extension": "^2.0",
        "behat/mink-browserkit-driver": "^1.2",
        "behat/mink-selenium2-driver": "^1.2",
        "dunglas/php-property-info": "^0.2",
        "phpdocumentor/reflection": "^1.0",
        "phpunit/phpunit": "^5.1",
        "jolicode/gif-exception-bundle": "^0.2"
    },
    "conflict": {
        "jms/serializer": "<=1.0.0",
        "knplabs/knp-menu": "<=2.0.1",
        "fzaninotto/faker": "<=1.5.0",
        "knplabs/knp-markdown-bundle": "1.5.0",
        "phpdocumentor/reflection": ">=1.1.0"
    },
    "scripts": {
        "post-root-package-install": [
            "SymfonyStandard\\Composer::hookRootPackageInstall"
        ],
        "post-install-cmd": [
            "./configure",
            "AppBundle\\Composer\\ScriptHandler::prepareParameters",
            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::removeSymfonyStandardFiles",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
        ],
        "post-update-cmd": [
            "./configure",
            "AppBundle\\Composer\\ScriptHandler::prepareParameters",
            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::removeSymfonyStandardFiles",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
        ]
    },
    "config": {
        "bin-dir": "bin"
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "extra": {
        "symfony-app-dir": "app",
        "symfony-web-dir": "web",
        "symfony-assets-install": "relative",
        "incenteev-parameters": {
            "file": "app/config/parameters.yml",
            "dist-file": "app/config/parameters-dist.yml"
        }
    }
}
```

**Comment by greg0ire** [comment]
> @greg0ire Prepare your heart attack. :heart: 

That's cute T___T

**Comment by andrey1s** [comment]
@Soullivaneuh 
under what conditions an error occurs(stack trace)? you use "sonata.admin.menu.matcher.voter.children" separately from "knp_menu.matcher"?
we can:
-  inject `service_container` in `sonata.admin.menu.matcher.voter.children` but its not best solution
- set `sonata.admin.menu.matcher.voter.children` private, but it still will not be able to inject other service
which solution will be better?

I was not able to repeat the error.
Only if I am directly appealing to the service, but I can not imagine what it might be needed.

**Pull request merged**